File: src/Test/PrimesFFI.js
export const runPrimesFFI = function(limit) {
  let n = Math.floor(limit);
  let count = 0;
  for (let i = 2; count < n; i++) {
    let isPrime = true;
    for (let j = 2; j * j <= i; j++) {
      if (i % j === 0) {
        isPrime = false;
        break;
      }
    }
    if (isPrime) {
      count++;
    }
  }
  return count;
};

File: src/Test/PrimesFFI.erl
-module(test_primesFFI@foreign).
-export([runPrimesFFI/1]).
divides(D, N) -> N rem D == 0.
any_divides(_N, { nil }) -> false;
any_divides(N, { cons, X, Xs }) ->
  case divides(X, N) of
    true -> true;
    false -> any_divides(N, Xs)
  end.
primes(Limit, Curr, Acc) when Curr > Limit -> Acc;
primes(Limit, Curr, Acc) ->
  case any_divides(Curr, Acc) of
    true -> primes(Limit, Curr + 1, Acc);
    false -> primes(Limit, Curr + 1, { cons, Curr, Acc })
  end.
sum_list({ nil }) -> 0;
sum_list({ cons, X, Xs }) -> X + sum_list(Xs).
runPrimesFFI(N) -> sum_list(primes(N, 2, { nil })).
